In [38]:
import joblib
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

In [39]:
df = pd.read_csv("../data/processed_data/feature_train.csv")

In [40]:
y = df["Churn"]
X = df.drop(columns = ["Churn"])

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2,stratify=y,random_state=42)

#### Load Saved Model From Models

In [42]:
lr = joblib.load("../models/logistic.pkl")
rf = joblib.load("../models/Random_forest.pkl")
xgb = joblib.load("../models/xgboost.pkl")

In [43]:
lr_pred = lr.predict(X_test)
rf_pred = rf.predict(X_test)
xgb_pred = xgb.predict(X_test)

#### Accuracy Scores For Each Models

In [44]:
print("LR Accuracy:", accuracy_score(y_test, lr_pred))
print("RF Accuracy:", accuracy_score(y_test, rf_pred))
print("XGB Accuracy:", accuracy_score(y_test, xgb_pred))

LR Accuracy: 0.8055358410220014
RF Accuracy: 0.7799858055358411
XGB Accuracy: 0.7806955287437899


#### Classification Report of Each Models

In [21]:
print("LR Report", classification_report(y_test, lr_pred))

LR Report               precision    recall  f1-score   support

           0       0.85      0.89      0.87      1035
           1       0.66      0.56      0.61       374

    accuracy                           0.81      1409
   macro avg       0.75      0.73      0.74      1409
weighted avg       0.80      0.81      0.80      1409



In [22]:
print("RF Report", classification_report(y_test,rf_pred))

RF Report               precision    recall  f1-score   support

           0       0.82      0.89      0.86      1035
           1       0.61      0.48      0.53       374

    accuracy                           0.78      1409
   macro avg       0.72      0.68      0.70      1409
weighted avg       0.77      0.78      0.77      1409



In [23]:
print("XGB Report", classification_report(y_test,xgb_pred))

XGB Report               precision    recall  f1-score   support

           0       0.84      0.87      0.85      1035
           1       0.60      0.53      0.56       374

    accuracy                           0.78      1409
   macro avg       0.72      0.70      0.71      1409
weighted avg       0.77      0.78      0.78      1409



#### Confusion Matrix

In [27]:
print("Logistic Regression")
print(confusion_matrix(y_test,lr_pred))

Logistic Regression
[[925 110]
 [164 210]]


# Model Evaluation Summary

Three machine learning models were evaluated:

- Logistic Regression
- Random Forest
- XGBoost

## Accuracy Scores

| Model | Accuracy |
|--------|----------|
| Logistic Regression | 80.55% |
| Random Forest | 77.99% |
| XGBoost | 78.07% |

Among all models, Logistic Regression achieved the highest accuracy.

---

## Logistic Regression Performance

### Classification Report Analysis

- Class 0 predictions performed very well with high precision and recall.
- Class 1 predictions were comparatively lower but still acceptable.
- Overall weighted F1-score was around 0.80.

### Confusion Matrix Analysis

```text id="o3yzqm"
[[925 110]
 [164 210]]

##### Although Logistic Regression achieved the highest accuracy, the recall score for churn prediction can still be improved. Further hyperparameter tuning and feature optimization may help improve churn detection performance.

In [28]:
results = pd.DataFrame({
    "Model": ["Logistic Regression", "Random Forest", "XGBoost"],
    "Accuracy": [0.8055, 0.7799, 0.7807]
})

results.to_csv("../reports/model_results.csv", index=False)

In [29]:
best_model_name = "Logistic Regression"

In [30]:
with open("../reports/best_model.txt", "w") as f:
    f.write(best_model_name)

In [35]:
report = report = classification_report(y_test, lr_pred)

In [36]:
with open("../reports/lr_classification_report.txt", "w") as f:
    f.write(report)

#### Best Model Among all of these

In [37]:
joblib.dump(lr, "../models/best_model.pkl")

['../models/best_model.pkl']